presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveUnsupportedDynamicFilters.java

```diff
@@ -0,0 +1,212 @@
+        public PlanWithConsumedDynamicFilters visitJoin(JoinNode node, Set<String> allowedDynamicFilterIds)
+        {
+            ImmutableSet<String> allowedDynamicFilterIdsProbeSide = ImmutableSet.<String>builder()
+                    .addAll(node.getDynamicFilters().keySet())
+                    .addAll(allowedDynamicFilterIds)
+                    .build();
+
+            PlanWithConsumedDynamicFilters leftResult = node.getLeft().accept(this, allowedDynamicFilterIdsProbeSide);
+            Set<String> consumedProbeSide = leftResult.getConsumedDynamicFilterIds();
+            Map<String, Symbol> dynamicFilters = node.getDynamicFilters().entrySet().stream()
+                    .filter(entry -> consumedProbeSide.contains(entry.getKey()))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            PlanWithConsumedDynamicFilters rightResult = node.getRight().accept(this, allowedDynamicFilterIds);
+            Set<String> consumed = new HashSet<>(rightResult.getConsumedDynamicFilterIds());
+            consumed.addAll(consumedProbeSide);
+            consumed.removeAll(dynamicFilters.keySet());
+
+            PlanNode left = leftResult.getNode();
+            PlanNode right = rightResult.getNode();
+            if (!left.equals(node.getLeft()) || !right.equals(node.getRight()) || !dynamicFilters.equals(node.getDynamicFilters())) {
+                return new PlanWithConsumedDynamicFilters(new JoinNode(
+                        node.getId(),
+                        node.getType(),
+                        left,
+                        right,
+                        node.getCriteria(),
+                        node.getOutputSymbols(),
+                        node.getFilter(),
+                        node.getLeftHashSymbol(),
+                        node.getRightHashSymbol(),
+                        node.getDistributionType(),
+                        node.isSpillable(),
+                        dynamicFilters), consumed);
+            }
+            return new PlanWithConsumedDynamicFilters(node, consumed);
+        }
+
+        @Override
+        public PlanWithConsumedDynamicFilters visitFilter(FilterNode node, Set<String> allowedDynamicFilterIds)
+        {
+            PlanWithConsumedDynamicFilters result = node.getSource().accept(this, allowedDynamicFilterIds);
+
+            Expression original = node.getPredicate();
+            Expression modified;
+            Set<String> consumedDynamicFilterIds = new HashSet<>(result.getConsumedDynamicFilterIds());
+
+            PlanNode source = result.getNode();
+            if (source instanceof TableScanNode) {
+                // Keep only allowed dynamic filters
+                modified = removeDynamicFilters(original, allowedDynamicFilterIds, consumedDynamicFilterIds);
+            }
+            else {
+                modified = removeAllDynamicFilters(original);
+            }
+
+            if (original.equals(modified)) {
```

Comment:
I think it will be simpler when to do new filter node condition here, e.g:
```
if (TRUE_LITERAL.equals(modified))
  return new PlanWithConsumedDynamicFilters(source, consumedDynamicFilterIds);
}

if (!original.equals(modified) || source != node.getSource()) {
  // create new filter node
}

// return same node
```